torchrec_dlrm/dlrm_main.py

```diff
@@ -499,75 +406,54 @@ class TrainValTestResults:
 
 def train_val_test(
     args: argparse.Namespace,
-    train_pipeline: TrainPipelineSparseDist,
+    pipeline: TrainPipelineSparseDist,
     train_dataloader: DataLoader,
     val_dataloader: DataLoader,
     test_dataloader: DataLoader,
-    lr_scheduler,
+    lr_scheduler: LRPolicyScheduler,
 ) -> TrainValTestResults:
     """
-    Train/validation/test loop. Contains customized logic to ensure each dataloader's
-    batches are used for the correct designated purpose (train, val, test). This logic
-    is necessary because TrainPipelineSparseDist buffers batches internally (so we
-    avoid batches designated for one purpose like training getting buffered and used for
-    another purpose like validation).
+    Train/validation/test loop.
 
     Args:
         args (argparse.Namespace): parsed command line args.
-        train_pipeline (TrainPipelineSparseDist): pipelined model.
-        train_dataloader (DataLoader): DataLoader used for training.
-        val_dataloader (DataLoader): DataLoader used for validation.
-        test_dataloader (DataLoader): DataLoader used for testing.
+        pipeline (TrainPipelineSparseDist): pipelined model.
+        train_dataloader (DataLoader): Training set's dataloader.
+        val_dataloader (DataLoader): Validation set's dataloader.
+        test_dataloader (DataLoader): Test set's dataloader.
+        lr_scheduler (LRPolicyScheduler): Scheduler to adjust the learning rate.
 
     Returns:
         TrainValTestResults.
     """
 
     train_val_test_results = TrainValTestResults()
 
-    train_iterator = iter(train_dataloader)
-    test_iterator = iter(test_dataloader)
     for epoch in range(args.epochs):
-        val_iterator = iter(val_dataloader)
+        new_pipeline = TrainPipelineSparseDist(pipeline._model, pipeline._optimizer, pipeline._device)
+        new_pipeline._context = pipeline._context
         _train(
-            train_pipeline,
-            train_iterator,
-            val_iterator,
+            new_pipeline,
+            train_dataloader,
             val_dataloader,
             epoch,
-            args.epochs,
             args.batch_size,
-            args.change_lr,
-            args.lr_change_point,
-            args.lr_after_change_point,
             lr_scheduler,
             args.print_lr,
             args.validation_freq_within_epoch,
             args.limit_train_batches,
             args.limit_val_batches,
         )
-        train_iterator = iter(train_dataloader)
-        val_next_iterator = (
-            test_iterator if epoch == args.epochs - 1 else train_iterator
-        )
-        val_accuracy, val_auroc = _evaluate(
-            args.limit_val_batches,
-            train_pipeline,
-            val_iterator,
-            val_next_iterator,
-            "val",
-        )
+        new_pipeline = TrainPipelineSparseDist(pipeline._model, pipeline._optimizer, pipeline._device)
```

Comment:
call this val_pipeline?

Reply:
The reason for the variable name `new_pipeline` is to communicate to viewers that the pipelines created for train, test, and validation are the same.  If we had names val_pipeline, train_pipeline, test_pipeline, I could see that raising the question in people's minds what's the difference between a val_pipeline and a test_pipeline, for example.

Reply:
A reasonable counter view is that val_pipeline, train_pipeline, test_pipeline would communicate which pipeline pairs with which dataloader, but I do not see a risk of mistaking one pipeline for another that would raise the need to create three pipeline variable names.